core/src/mindustry/ui/dialogs/ResearchDialog.java

```diff
@@ -792,4 +859,43 @@ public void drawChildren(){
             super.drawChildren();
         }
     }
-}
+    //Whether this player is allowed to research (for use in multiplayer campaign)
+    public boolean researchAllowed(@Nullable Player player){
+        return !Vars.net.active() || (player != null && (player.admin() || !state.rules.researchRequiresAdmin));
+    }
+    // TODO shouldn't this have a target?
+    @Remote
+    public static void partiallyResearched(Content content, int[] ids, int[] quantity){
+        if(!(content instanceof UnlockableContent u) || u.techNode == null) return;
+        if(net.client()) { // TODO isn't this always true?
+            Events.fire(new PartialResearchEvent(u, ids, quantity));
+        }
+    }
+    @Remote(targets = Loc.client)
+    public static void clientResearch(@Nullable Player player, Content content){
+        if(!ui.research.researchAllowed(player)) return;
+        if(!(content instanceof UnlockableContent u) || !ui.research.view.canSpend(u.techNode)) return;
+        ui.research.rebuildItems();
+        ui.research.view.spend(u.techNode);
+    }
+    @Remote(targets = Loc.client)
+    public static void requestPlanetItems(@Nullable Player player) {
+        if(!ui.research.researchAllowed(player)) return;
+        int[] quantity = new int[Vars.content.items().size];
+        ui.research.rebuildItems();
+        int index = 0;
+        for (Item item : Vars.content.items()) {
+            quantity[index++] = ui.research.items.get(item);
+        }
+        Call.sendPlanetItems(player.con, quantity);
+    }
+    @Remote(targets = Loc.server, variants = Variant.one)
```

Comment:
```suggestion
    @Remote(targets = Loc.server)
```
You should be able to send planetItems to more than 1 client at a time

Reply:
They won't notice it anyway.

Reply:
> prevents desync if everyone is updated same time

the dialog takes time to open, it should be long enough for the item displays/partial research progress to update.

If your ping is extremely high, the packets can arrive very late. Then, the solution is to update all clients periodically (every X milliseconds or such). But I actually didn't learn arc.util.Timer and don't have a means to test high latency, so I won't be able to do this.